### ISSUE TITLE
Subsetting sce

### ISSUE THREAD

**Issue by MartijnVNL** [HelenaLC/CATALYST#166]
Hi Helena,

I'm having a lot of fun with Catalyst, but I ran into an issue that I can't seem to solve. Currently I'm using a large dataset containing both adults and children, with a lot of different conditions and timepoints. I've put all that into the metadata file and created the single cell experiment with prepData, which all works fine. But now I would like to do some subsetting. So in some cases I want to compare adults in children, but in other cases I want to compare conditions within adults only. That means I want to subset the children out. So let's say I have performed FlowSOM on sce (containing both children and adult data) and now I want to plot the metacluster abundance of only the adults. This is my code ("Age_class" contains either "Adult", or "Child"):

`metClust = 12
meta = gsub(" ", "", paste("meta", metClust))
sce <- cluster(sce, features = type_markers(sce), xdim = 10, ydim=10, maxK=metClust, seed=1234) 
plotAbundances(sce[,sce$Age_class=="Adult"], k = meta, by = "cluster_id", group_by = "sample_id")`

It then plots the data, but it gives the following error:

![image](https://user-images.githubusercontent.com/73521792/101467215-53b61a00-3942-11eb-8183-5042cd98fa06.png)

I know this is a ggplot error, but I can't find what's causing it. My guess is that there is something wrong with the way I subset, because when I import only the Adult files I don't get an error. However, when I call sce of either the subsetted Adults, or when I only import the fcs files of the Adults and compare the little summaries, they look the same. And when I compare the sce summary of the subsetted Adults to the non-subsetted total, the subsetted sce has less events in "dim", suggesting that the subsetting worked.

So my question is if I'm subsetting right and if so how I can fix the ggplot error. If I'm not subsetting right, is there a way to do it properly?

Kind regards,

Martijn

**Comment by HelenaLC** [comment]
Yes, there is a way. We provide the `filterSCE()` function for this - mainly because 
- basic subsetting such as `sce[, variable != ...]` does not affect the SCE's metadata, and will retain no-longer existent factor levels.
- In addition, `filterSCE()` allows some `CATALYST`-specific filtering, specifically, filtering on different clusterings (e.g. `filterSCE(sce, k = "meta10", cluster_id %in% c(1,5,8))`). 

The basic syntax is a la `dplyr::filter()`, so you'd want to do something like `filterSCE(sce, Age_class == "Adult")`

**Comment by MartijnVNL** [comment]
Unfortunately it doesn't seem to be working. I'm trying to plot metacluster abundances like this:

`metClust = 25`
`meta = gsub(" ", "", paste("meta", metClust)`

`sce <- cluster(sce, features = type_markers(sce), xdim = 10, ydim=10, maxK=metClust, seed=1234)`
`plotAbundances(filterSCE(sce, Age_class == "Adult"), k = meta, by = "cluster_id", group_by = "condition")`

But I get an error:

`Error in `[[<-.data.frame`(`*tmp*`, i, value = integer(0)) : replacement has 0 rows, data has 25`

**Comment by HelenaLC** [comment]
1. Could you post the output of `show(sce)`, `head(colData(sce))` as well as `names(cluster_codes(sce))`?
2. FYI: You can use `paste0` to paste without spaces instead of `gsub(" ", ...)` ;)

**Comment by MartijnVNL** [comment]
Certainly! Because I couldn't subset I also put the Age_class and Timepoint data in the condition column of the metadata for now. Thanks for the tip btw. I'm still pretty new to R, and I had to skip over the basics.

metClust = 15 ##set number of metaclusters
meta = paste0("meta", metClust))##creates variable equal to amount of metaclusters for k=meta

`show(sce)`
`class: SingleCellExperiment 
dim: 13 4830612 
metadata(4): experiment_info cluster_codes SOM_codes delta_area
assays(2): counts exprs
rownames(13): CD57 CD27 ... CCR4 CD28
rowData names(4): channel_name marker_name marker_class used_for_clustering
colnames: NULL
colData names(10): sample_id condition ... Batch cluster_id
reducedDimNames(0):
altExpNames(0):`

`head(colData(sce))`
`DataFrame with 6 rows and 10 columns
      sample_id         condition patient_id      Age Age_group Age_class   Gender Timepoint
       <factor>          <factor>   <factor> <factor>  <factor>  <factor> <factor>  <factor>
1 449002_S+_M_A Symptoms+_Adult_1     449002       42     35-49     Adult     Male         1
2 449002_S+_M_A Symptoms+_Adult_1     449002       42     35-49     Adult     Male         1
3 449002_S+_M_A Symptoms+_Adult_1     449002       42     35-49     Adult     Male         1
4 449002_S+_M_A Symptoms+_Adult_1     449002       42     35-49     Adult     Male         1
5 449002_S+_M_A Symptoms+_Adult_1     449002       42     35-49     Adult     Male         1
6 449002_S+_M_A Symptoms+_Adult_1     449002       42     35-49     Adult     Male         1
     Batch cluster_id
  <factor>   <factor>
1       NA         80
2       NA         58
3       NA         1 
4       NA         79
5       NA         12
6       NA         61`

`names(cluster_codes(sce))`
`[1] "som100" "meta2"  "meta3"  "meta4"  "meta5"  "meta6"  "meta7"  "meta8"  "meta9"  "meta10"
[11] "meta11" "meta12" "meta13" "meta14" "meta15"`

**Comment by HelenaLC** [comment]
1. Shouldn't the 2nd group of statements be `filterSCE(sce, Age_class == "Adult")` (additional quotes, no "s")?
2. Could you also, with *all* issues, post the output of your `sessionInfo()` so we can rule out R/package-version related causes.

**Comment by MartijnVNL** [comment]
It should indeed be `filterSCE(sce, Age_class == "Adult")`, but that was a typo I made when posting this. It is correct in R. I couldn't copy paste because I was running R on a virtual desktop via Citrix and GitHub on my own desktop. I ran it again just to be safe:

`show(sce)`
`cat("\n")`
`head(colData(sce))`
cat("\n")
`names(cluster_codes(sce))`

![image](https://user-images.githubusercontent.com/73521792/102482505-9626d780-4063-11eb-8f26-71424acea75d.png)

show(filterSCE(sce, Age_class == "Adult"))
cat("\n")
head(colData(filterSCE(sce, Age_class == "Adult")))
cat("\n")
names(cluster_codes(filterSCE(sce, Age_class == "Adult")))

![image](https://user-images.githubusercontent.com/73521792/102482619-c66e7600-4063-11eb-8096-f044242af00c.png)

As for the errors:

In the condition column of the metadata there are 12 levels. When I plot the unsubsetted data it only shows 7 levels. In addition, for example the Hospitalized_Adult_2 condition has 7 datapoints if I look in Excel, but only 1 is shown in the plot.
`unique(md$condition)`

![image](https://user-images.githubusercontent.com/73521792/102484515-ad1af900-4066-11eb-91d2-507db00ec2bb.png)

`plotAbundances(filterSCE(sce, Age_class == "Adult"), k = meta, by = "cluster_id", shape = "Gender")`

![image](https://user-images.githubusercontent.com/73521792/102485581-1cddb380-4068-11eb-81e0-50f595a279cd.png)

If I then subset the data I get an error:

`plotAbundances(filterSCE(sce, Age_class == "Adult"), k = meta, by = "cluster_id", shape = "Gender")`

![image](https://user-images.githubusercontent.com/73521792/102485858-76de7900-4068-11eb-9b1b-0b32571906da.png)

`sessionInfo()`

![image](https://user-images.githubusercontent.com/73521792/102486002-aee5bc00-4068-11eb-928f-8a8bbacf07db.png)

**Comment by HelenaLC** [comment]
Hm. Could you post something like table(sce$condition, sce$gender) and table(condition, age_class) etc? (Before filtering)

**Comment by MartijnVNL** [comment]
Sure! I didn't include sample_id, patient_id and age, because there are a couple hundred donors, so the table will be massive.

`table(sce$condition, sce$Age_group)`

![image](https://user-images.githubusercontent.com/73521792/102500454-a0090480-407c-11eb-88b6-424b394bda56.png)


table(sce$condition, sce$Age_class)

![image](https://user-images.githubusercontent.com/73521792/102500500-abf4c680-407c-11eb-8e90-aa4b20b77740.png)


table(sce$condition, sce$Gender)

![image](https://user-images.githubusercontent.com/73521792/102500542-b616c500-407c-11eb-8edc-c8e99bcc7ab8.png)


table(sce$condition, sce$Timepoint)

![image](https://user-images.githubusercontent.com/73521792/102500576-c038c380-407c-11eb-84d9-92ba20eb769f.png)

**Comment by HelenaLC** [comment]
Okay, I still don't know what's going on. However, I just saw in the screenshot you are running `CATALYST` v1.12.0. The current version is 1.14, and there were quite significant updates & bug fixes this past release. So I suggest updating via
```
BiocManager::install("CATALYST")
```
And, in general, checking frequently that your overall package versions are up-to-date and match each other 
```
BiocManager::version() # should be 3.12
BiocManager::valid()   # will show any discrepancies
BiocManager::install() # to update all packages to their current release version
```
*side note: Best to post the output of `sessionInfo()` as a code chunk rather than screenshot ;)

**Comment by MartijnVNL** [comment]
Sorry for the late reply, I was on Christmas break. I finally managed to fix the problem. I first updated everything, but the problem persisted. Because some of the conditions were not showing in the legends I thought the problem might be in the metadata file. I then noticed that the SampleID of every donor was the same between timepoints. After making sure that every file had a unique SampleID and making sure there were no "NA" cells in the metadata file everything worked flawlessly.

Thanks for all the help!
